### PR TITLE
refactor globalhub status

### PR DESCRIPTION
**Pull request by ldpliu**
## Summary
```
[root@dangpeng ~]# oc get mgh
NAME                    STATUS    AGE
multiclusterglobalhub   Running   20h
```

```
  status:
    components:
      kafka:
        kind: Kafka
        lastTransitionTime: "2024-09-24T00:14:31Z"
        message: Kafka cluster is ready
        name: kafka
        reason: KafkaClusterReady
        status: "True"
        type: Available
      multicluster-global-hub-grafana:
        kind: Deployment
        lastTransitionTime: "2024-09-23T05:30:12Z"
        message: Component multicluster-global-hub-grafana have been deployed
        name: multicluster-global-hub-grafana
        reason: MinimumReplicasAvailable
        status: "True"
        type: Available
      multicluster-global-hub-manager:
        kind: Deployment
        lastTransitionTime: "2024-09-24T00:14:33Z"
        message: Component multicluster-global-hub-manager have been deployed
        name: multicluster-global-hub-manager
        reason: MinimumReplicasAvailable
        status: "True"
        type: Available
      multicluster-global-hub-postgres:
        kind: StatefulSet
        lastTransitionTime: "2024-09-23T05:28:32Z"
        message: Component multicluster-global-hub-postgres have been deployed
        name: multicluster-global-hub-postgres
        reason: MinimumReplicasAvailable
        status: "True"
        type: Available
    conditions:
    - lastTransitionTime: "2024-09-23T05:28:14Z"
      message: The data will be kept in the database for 18 months.
      reason: DataRetentionParsed
      status: "True"
      type: Database
    - lastTransitionTime: "2024-09-24T00:14:33Z"
      message: Multicluster Global Hub is ready, All the components avaible
      reason: MulticlusterGlobalHubReady
      status: "True"
      type: Ready
    - lastTransitionTime: "2024-09-23T05:28:14Z"
      message: Backup Disabled In RHACM
      reason: BackupDisabled
      status: "False"
      type: BackupLabelAdded
    phase: Running
```
## Related issue(s)
https://issues.redhat.com/browse/ACM-13680
Fixes #

## Tests
* [x] Unit/function tests have been added and incorporated into `make unit-tests`.
* [x] Integration tests have been added and incorporated into `make integration-test`.
* [x] E2E tests have been added and incorporated into `make e2e-test-all`.
* [ ] List other manual tests you have done.

### CODE REVIEW

File: operator/pkg/controllers/backup/backup_reconcile.go [hunk 1]
```diff
@@ -126,14 +132,18 @@ func addBackupCondition(ctx context.Context, client client.Client,
 	if err != nil {
 		msg = fmt.Sprintf("Added backup labels with error: %v.", err.Error())
 	}
-	if err := config.SetCondition(ctx, client, mgh,
-		config.CONDITION_TYPE_BACKUP,
-		config.CONDITION_STATUS_TRUE,
-		config.CONDITION_REASON_BACKUP,
-		msg,
-	); err != nil {
+	if err := config.UpdateCondition(ctx, client, types.NamespacedName{
```
Comment (clyang82) [inline operator/pkg/controllers/backup/backup_reconcile.go:135]: Can it be covered in statusReconciler?

File: operator/pkg/controllers/backup/backup_reconcile.go [hunk 2]
```diff
@@ -108,14 +110,18 @@ func addDisableCondition(ctx context.Context, client client.Client,
 	if err != nil {
 		msg = fmt.Sprintf("Backup is diabled with error: %v.", err.Error())
 	}
-	if err := config.SetCondition(ctx, client, mgh,
-		config.CONDITION_TYPE_BACKUP,
-		config.CONDITION_STATUS_FALSE,
-		config.CONDITION_REASON_BACKUP_DISABLED,
-		msg,
-	); err != nil {
+	if err := config.UpdateCondition(ctx, client, types.NamespacedName{
```
Comment (clyang82) [inline operator/pkg/controllers/backup/backup_reconcile.go:113]: Can it be covered in statusReconciler?

File: operator/pkg/config/condition.go [hunk 3]
```diff
@@ -86,11 +88,15 @@ const (
...
 const (
-	CONDITION_TYPE_GLOBALHUB_READY    = "Ready"
-	CONDITION_REASON_GLOBALHUB_READY  = "MulticlusterGlobalHubReady"
-	CONDITION_MESSAGE_GLOBALHUB_READY = "Multicluster Global Hub is ready"
-	CONDITION_REASON_GLOBALHUB_FAILED = "MulticlusterGlobalHubFailed"
-	CONDITION_REASON_GLOBALHUB_PAUSED = "MulticlusterGlobalHubPaused"
+	CONDITION_TYPE_GLOBALHUB_READY        = "Ready"
+	CONDITION_REASON_GLOBALHUB_READY      = "MulticlusterGlobalHubReady"
+	CONDITION_REASON_GLOBALHUB_NOT_READY  = "MulticlusterGlobalHubNotReady"
+	CONDITION_MESSAGE_GLOBALHUB_READY     = "Multicluster Global Hub is ready, All the components avaible"
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/config/condition.go]: Done

File: operator/pkg/config/condition.go [hunk 4]
```diff
@@ -86,11 +88,15 @@ const (
...
+	CONDITION_MESSAGE_GLOBALHUB_NOT_READY = "Multicluster Global Hub is not ready, Waiting all the components avaible"
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/config/condition.go]: Done

File: operator/pkg/config/condition.go [hunk 5]
```diff
@@ -86,11 +88,15 @@ const (
...
+
+	CONDITION_REASON_GLOBALHUB_FAILED     = "MulticlusterGlobalHubFailed"
+	CONDITION_REASON_GLOBALHUB_UNINSTALL  = "MulticlusterGlobalUninstalling"
+	CONDITION_MESSAGE_GLOBALHUB_UNINSTALL = "Multicluster Global Hub is Uninstalling"
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/config/condition.go]: Done

File: operator/pkg/controllers/hubofhubs/controller.go [hunk 6]
```diff
@@ -608,15 +607,39 @@ func (r *GlobalHubReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 	r.log.V(2).Info("reconciling mgh instance", "namespace", req.Namespace, "name", req.Name)
 	mgh, err := config.GetMulticlusterGlobalHub(ctx, r.client)
 	if err != nil {
+		if errors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/controllers/hubofhubs/controller.go]: Done

File: operator/pkg/controllers/status/status_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+		err = r.updateMghStatus(ctx, componentsStatus)
+		if err != nil {
+			klog.Errorf("Failed to update mgh status, err: %v", err)
+		}
+	}()
+
+	// update manager/grafana/inventory api
+	err = updateDeploymentComponents(ctx, r.Client, mgh.Namespace, componentsStatus)
+	if err != nil {
+		klog.Errorf("Failed to update deployment components status:%v", err)
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/controllers/status/status_reconciler.go]: Done

File: operator/pkg/controllers/status/status_reconciler.go [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+		}
+		return
+	}
+	componentsStatus[name] = v1alpha4.StatusCondition{
+		Kind:               resourceType,
+		Name:               name,
+		Type:               config.COMPONENTS_AVAILABLE,
+		Status:             config.CONDITION_STATUS_FALSE,
+		Reason:             config.MINIMUM_REPLICAS_UNAVAILABLE,
+		Message:            fmt.Sprintf("Component %s have been deployed, but not ready", name),
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply operator/pkg/controllers/status/status_reconciler.go]: Done

File: operator/pkg/controllers/backup/backup_reconcile.go (hunk 1 above)
Comment (ldpliu) [reply operator/pkg/controllers/backup/backup_reconcile.go:135]: It's difficult to move it to statusreconciler, as we need to cache backup reconcile error in this controller.

**Review by clyang82 — APPROVED** [review]
LGTM

**Pull request merged**